src/concrete/ml/quantization/quantized_ops.py

```diff
@@ -217,20 +222,113 @@ def q_impl(
+                assert (
+                    batch_a == batch_b or batch_a == 1 or batch_b == 1
+                ), "Batch sizes must be equal or one must be 1"
+
+                batch_size = batch_a
+                c = zeros(shape=(batch_size, m, p))
+
+                for i in range(batch_size):
+                    a_slice = a[i] if a_3d else a
+                    b_slice = b[i] if b_3d else b
+
+                    # Reshape for element-wise multiplication
+                    a_reshaped = a_slice.reshape((m, n, 1))
+                    b_reshaped = b_slice.reshape((1, n, p))
+                    enc_mul_result = enc_mul(a_reshaped, b_reshaped)
+                    c[i] = numpy.sum(enc_mul_result, axis=1)
+
+                # If both inputs were 2D, remove the first dimension
+                if not a_3d and not b_3d:
+                    c = numpy.squeeze(c, axis=0)
+
+                return c
+
+        # FIXME: https://github.com/zama-ai/concrete-ml-internal/issues/4127
+        @univariate
+        def copy_function(x):
+            return x
+
+        # This is done to block the precision raising at compilation time
+        # Here we add a PBS to protect inputs precision. We don't want to do it
+        # when it is not needed as it is costly. We just use it for encrypted matmul
+        input_q_values_matmul = (
+            copy_function(input_q_values) if is_encrypted_gemm else input_q_values
+        )
+        input2_q_values_matmul = (
+            copy_function(input2_q_values) if is_encrypted_gemm else input2_q_values
+        )
 
         # Core matmul operation in full integers with a shape change (INTEGERS)
         with tag(self.op_instance_name + ".matmul"):
-            matmul = input_q_values @ weights_q_values
+            # We implement our own encrypted matmul to be able to round before PBS
+            if is_encrypted_gemm:
+                matmul = matmul(input_q_values_matmul, input2_q_values_matmul)
+            # Otherwise we let concrete do it
+            else:
+                matmul = input_q_values_matmul @ input2_q_values_matmul
+
+        input_q_values_sum = copy_function(input_q_values) if is_encrypted_gemm else input_q_values
 
         # If the weights have symmetric quantization, their zero point will be 0
         # The following check avoids the computation of the sum of the inputs, which may have
         # large bit-width, in the case where it would be multiplied by zero
-        if q_weights.quantizer.zero_point != 0:
+        if q_input2.quantizer.zero_point != 0:
             # Sum operation in full integers resulting in large integers (INTEGERS)
             with tag(self.op_instance_name + ".matmul_inputsum"):
-                sum_input = -q_weights.quantizer.zero_point * numpy.sum(
-                    input_q_values, axis=1, keepdims=True
+                sum_input = -q_input2.quantizer.zero_point * numpy.sum(
+                    input_q_values_sum, axis=-1, keepdims=True
```

Comment:
is this `axis=-1` change expected ?

Reply:
Yes. The sum should happen on the last axis. In 2d it was 1 but that doesn't generalize to 3d. -1 does.